hal/src/dmac/transfer.rs

```diff
@@ -286,21 +285,24 @@ impl<B: AnyBufferPair> AsMut<B> for SpecificBufferPair<B> {
 // TODO change source and dest types to Pin? (see https://docs.rust-embedded.org/embedonomicon/dma.html#immovable-buffers)
 /// DMA transfer, owning the resources until the transfer is done and
 /// [`Transfer::wait`] is called.
-pub struct Transfer<Chan, Buf, Pld = ()>
+pub struct Transfer<Chan, Buf, Pld = (), W = fn()>
 where
     Buf: AnyBufferPair,
     Chan: AnyChannel,
+    W: FnMut() + 'static,
 {
     chan: Chan,
     buffers: Buf,
     payload: Pld,
+    waker: Option<W>,
```

Comment:
I just had an idea. With this approach, you always pay for the space of a function pointer in every transfer struct. Maybe you could remove the Fn() bound on the struct and set the default type to (). You can apply the bound to the with_waker function and any function where it would be used. I think that would work.